Add plugin support and refactor render graph.

freezy opened the pull request:
The PR adds external plugin support to dmdext, originally provided by @lucky01 in #347.

I've chosen a more generic approach. Instead of hard-coded DLL names, users will need to enable the plugin in DmdDevice.ini. For pin2color, it would look something like that:

```ini
; load plugins
plugin.0.path = C:\Visual Pinball\VPinMAME\pin2color.dll
plugin.0.path64 = C:\Visual Pinball\VPinMAME\pin2color64.dll
plugin.0.passthrough = true
```

The `0`, because, well, there might be other plugins in the future.

The biggest chunk of this PR however is integration of #239, although I didn't tackle the multi resolution feature yet. But the render graph now exclusively deals with frame objects, which is much cleaner, less code and more comprehensible. Also, there are now unit tests covering the render graph routing and the frame conversion and transformation code.

The reason for getting #239 into the plugin branch was because the transformation logic was a mess, and I ended up not knowing the relation between the dimensions of a frame and the data of the frame, since they were de-coupled.

Now, every frame's dimensions *always* correlate with the data. Usually, the transformation logic is the following. Given a colored frame:

1. If HD scale is enabled and destination supports 2x resolution, upscale.
2. Convert to destination's frame format
3. Downscale or center, if fixed-size destination.

This is still WIP, lots of stuff can be cleaned up and improved, but it's functional. I'll spend the next days integrating the other fixes of #347, and then we could start some public beta testing.

@lucky01 let me know if this works for you. I couldn't test any PAC files because I don't have pin2color.dll with PAC support.

This PR closes #347, #239, #7, #323, #380, #330, #234.

lucky01 commented:
I´m currently on vacation and can start testing earliest in 3 weeks . 
From the plugin side it makes no difference wether you have pac / vni or fsq as a source . If it works with one on your side it should work with all 3 . Did you test all source resolutions 128x32 128x16 192x64 and 256x64 ? With and without scaling ?
Is the plugin working with all of that ?
BTW there is no such thing like Gray6Source. No pinball machine sends frames in 6bit format. That can be removed from the code. It is a relict from the first days when we started to implement colored gray 6 
6bit is always colored gray 6.

freezy commented:
> From the plugin side it makes no difference wether you have pac / vni or fsq as a source . If it works with one on your side it should work with all 3 .

Yes, the problem is just that my VNI collection is somewhat limited, and PAC files is what people will finally use.

If you could give me a few non-standard game names, that would help. Specially games with 128x16 input and games with HD output. Then I can hunt down those VNIs. Also, you were mentioning segdisplay colorizations?

> Did you test all source resolutions 128x32 128x16 192x64 and 256x64 ? With and without scaling ?
Is the plugin working with all of that ?

A little bit in the train, and it crashed. ;) Looking into it.

> BTW there is no such thing like Gray6Source.

Yes, I figured. Will clean that up.

Have a good vacation!

lucky01 commented:
For 128x16 there are hook, batman, star trek 25th, checkpoint and tmnt. Games with 256x64 are Capcom Football Flipper but there is no colorization. Seg Display there is Lupin and Ken based on diner both presented by darden78 on vpu, but I can send you my testing project when I´m back.

freezy commented:
> Seg Display there is Lupin and Ken based on diner both presented by darden78 on vpu, but I can send you my testing project when I´m back.

Sorry, I missed this. I'll ping darden78 on VPU, maybe he's willing to share.

lucky01 commented:
> > Seg Display there is Lupin and Ken based on diner both presented by darden78 on vpu, but I can send you my testing project when I´m back.
> 
> Sorry, I missed this. I'll ping darden78 on VPU, maybe he's willing to share.

I sent you a simple test project

freezy commented:
Okay, I'll have a look. Right now, when rendering the rotations for the virtual DMD, we actually only update the palette in the shader (as opposed to rendering an entire RGB24 texture).

lucky01 commented:
Your LibDmd/DmdDevice/AlphaNumeric.cs does also not have my latest fixes

freezy commented:
> Your LibDmd/DmdDevice/AlphaNumeric.cs does also not have my latest fixes

Hmm, I'm pretty sure I did pull your changes. I haven't gone through all of them yet, but this one seems up to date.

[Your version](https://github.com/freezy/dmd-extensions/blob/a7961e475adf12051eac91caff5afc93a1348456/LibDmd/DmdDevice/AlphaNumeric.cs) versus [mine](https://github.com/freezy/dmd-extensions/blob/b0f1e16222f1b81fe55e417a10d907d5168a056a/LibDmd/DmdDevice/AlphaNumeric.cs). 

Anything particular that is missing?

lucky01 commented:
I just had a look at your plugin code. Why do you try to split the RGB24 frame in palette based frames in there, while the majority of the supported devices does not need that ? Why not only split frames and generate a palette when the output device needs it to overcome performance lags that way ? So far I have no performance issues reported from pindmdv3, pixelcade or pin2dmd users processing the colorized frames as RGB24.

freezy commented:
Because it's the frame format that goes into the render graph, and once it's in there, subscribers don't know whether it's a frame that can be converted back to colored gray. I profiled it, it turned out not to be an expensive operation.

lucky01 commented:
Did you take both the splitting into palette based and the conversion back to RGB24 for e.g. the virtual dmd into account for your profiling ? I introduced a coloredgray source to be able to differentiate a normal RGB24 frame from a colorized RGB24 frame in the rendergraph.

freezy commented:
The virtual DMD renders colored gray frames, the RGB conversion is done in the shader.

It's funny that you seem to be concerned about performance now, when you could simply return colored gray frames in the plugin and get the best of both worlds. ;)

lucky01 commented:
The plugin it attached to the 15ms frame timing of pinmame

freezy commented:
What happens if the plugin only gets frames when they change? Some animations don't play?

lucky01 commented:
I expect some scenes have a wrong timing e.g LAH with movie replacement scenes,
They will not be updated depending on the scene frame timing but on the source change timing.
As a sidenote I tested frame deduplication to the pinup output and some pup frames did not trigger that way.

lucky01 commented:
https://github.com/freezy/dmd-extensions/pull/347#issuecomment-1326797186

freezy commented:
Okay, thanks.

I'll de-dupe after colorization then. TBH I have never used PinUP and @nailbuster is currently on holidays.

I'll release a beta version of this branch tonight, since PinballFX is launching a new version today with dmddevice support, for which I had to add a few fixes, which ended up in this branch.

Hopefully I'll get the de-duping fixed before, otherwise that will be another beta.

Are you back from holidays and did you happen to test this a bit more?

lucky01 commented:
No ! Back on the 19thAm 08.06.2023 um 19:32 schrieb freezy ***@***.***>:﻿
Okay, thanks.
I'll de-dupe after colorization then. TBH I have never used PinUP and @nailbuster is currently on holidays.
I'll release a beta version of this branch tonight, since PinballFX is launching a new version today with dmddevice support, for which I had to add a few fixes, which ended up in this branch.
Hopefully I'll get the de-duping fixed before, otherwise that will be another beta.
Are you back from holidays and did you happen to test this a bit more?

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

lucky01 commented:
> Okay, I'll have a look. Right now, when rendering the rotations for the virtual DMD, we actually only update the palette in the shader (as opposed to rendering an entire RGB24 texture).

Does color rotation work for all output devices now like in my version ?

freezy commented:
No, but I've added back frames at 60Hz when going through the coloring plugin.

lucky01 commented:
> No, but I've added back frames at 60Hz when going through the coloring plugin.

O.K. then I will wait until your finished haqving a look at it.

freezy commented:
Last two commits should enable support for color rotation on all devices. It's a new provider that emits palette changes, triggered by Serum. In the Pin2DMD code, I've just made it run `SetSinglePalette()` with the new palette. Let me know if that works.

lucky01 commented:
I saw that you renamed the deprecated colorizer code into pin2color. To avoid confusion with the plugin could you please rename it to deprecated as that is what it is.

freezy commented:
Hmm, I don't see the confusion.

- Pin2Color -> The Pin2DMD colorization system. 
- Plugin -> The plugin system.

As you might have noticed, nothing in the plugin system is Pin2Color-specific, and that's on purpose. So the naming makes sense for me...

Could you test the color rotations on PIN2DMD?

lucky01 commented:
> Hmm, I don't see the confusion.

It is not clear that way when users talk about pin2color wether they mean the old colorizer code or the new plugin. 

Please rename it ! 

I´m just looking at the rotation code because it is not working that way.

lucky01 commented:
UpdatePalette needs to be like this to work with pin2dmd

public void UpdatePalette(Color[] palette)
{
	SetSinglePalette(palette);

	// send frame buffer to device
	RenderRaw(_frameBufferGray6);
}

What about the other devices like pixelcade, pindmdv3, avi export etc ?

lucky01 commented:
@freezy
fix: Don't send 3-bit frames emitted by PinMAME.

Why this ?

freezy commented:
Well, this is in `InternalRenderRaw2Device()`, so 4-bit frames will go through `InternalRenderRaw4Device()`, no? Reason for this was because I was getting three bit planes, which didn't pass the internal checks I'm doing for 2-bit frames.

lucky01 commented:
> Well, this is in `InternalRenderRaw2Device()`, so 4-bit frames will go through `InternalRenderRaw4Device()`, no?

No, they come in on RenderRaw2 because the calculated result of the rawframes is a 2bit frame with 4 shades.

freezy commented:
Sorry I'm just trying to understand this. You say

> No, they come in on RenderRaw2 because the calculated result of the rawframes is a 2bit frame with 4 shades.

To be clear, 4 shades = 2-bit. So, why "no"?

My other questions:

- What do you mean by "calculated result"?
- What should I do with 3-bit frames then? Expand them to 4-bit?

lucky01 commented:
e.g. WPC has 4 shades in currbuffer.  noOfRawFrames for WPC is 3 and those 3 rawframes are in currrawbuffer.
All data coming from pinmame needs to be passed as it is without any manipulation to my plugin.

freezy commented:
Okay, so let me rephrase my question: Why are there 3 bit planes (which equal to 8 shades), when there are only 4 shades in WPC? What's in the third plane?

lucky01 commented:
To avoid misunderstandings and to save time just have a look at the pinmame source to see what is in rawframe buffer.
From the plugin side all I can say is that I need the pinmame data passed through to my plugin without modification for dumping etc.

freezy commented:
So you don't know what's in the third plane but you know the plugin needs it? Doesn't make much sense tbh..

lucky01 commented:
Of course I know it but just not the nerve and time to explain the priciple of pinmame dmdframes. Rawframes have been introduced several years ago and until now you don´t know what they are for ? Doesn´t make much sense either.

freezy commented:
Look, I'm trying. I know what raw frames are, but to my knowledge they come either in 2 or 4 bits. Now I have 3-bit planes, and the bit length in dmdext is quite strict, to make debugging easier.

I'm not asking in order to purposefully annoy you, and you should know me long enough to have noticed that I won't implement stuff that I don't understand just because you tell me to.

I can do some tests later to analyze the third bit planes, but if they are empty I'll keep the code as-is, since they are of no purpose, unless you're able to explain it to me.

lucky01 commented:
The rawframebuffer contains the data for a frame like sent to the plasma display on a real pinball machine.

freezy commented:
That doesn't explain anything but further adds confusion, since the plugin in dmdext doesn't run on real machines.

freezy commented:
> Funny, because you implemented that interface years ago without a single question
@vbousquet and @djrobx implemented that, as you can easily see with a git blame.

> but maybe your interest has grown since serum
Serum has nothing to do with that, I did a big refactor because of the plugin, and I want to understand how this works.

The alphanum stuff came from @zesinger as already noted above (#377), I just integrated it properly.

It's just sad how unwilling you are to cooperate on this. This started as a purely technical discussion, yet you somehow managed to pull this around into a political vengeance.

lucky01 commented:
That is why we agreed to have a plugin API. To avoid such discussions. You forward the pinmame data and I return the RGB24 frame. That´s it.

freezy commented:
Yes, but it's not done yet. To get there, we need communication, and you're making it really hard right now.

lucky01 commented:
> I'll be maintaining the code base, so please let me decide for myself how I name my classes.

... and then you use a name I came up with for my plugin ? Why not name it VNIColorizer or anything else ?

freezy commented:
Well, it sounded like a good term for your colorization algorithm. But if that's really so disturbing to you, I can rename it to `VniColorizer` as well.

freezy commented:
Okay I've reverted the bit length check and renamed the VNI colorizer.

One thing I've noticed is that the plugin doesn't take in the scaler mode for `Advanced256x64`. It seems to always use doubler. In the native VNI colorer this is configurable. Do you want to add this config option to the plugin as well?

lucky01 commented:
> One thing I've noticed is that the plugin doesn't take in the scaler mode for `Advanced256x64`.

In Advanced256x64 the frames are returned as 256x64 and no scaling is taking place. So no need for a scaling option for the plugin. The scaling is only done when the plugin returns 128x32 frames.

freezy commented:
You've sent me a test version of MM which contains frames at 256x64. The plugin returns all frames at that resolution. Thus, the plugin does the up-scaling for frames that aren't colored at 256x64. If you don't want to offer an option how to up-scale those frames, fine, but in the native VNI colorizer you can choose which algorithm to use.

lucky01 commented:
Like you said it is a test project which only contians a few colorized scenes. A final version of a colorization should have all frames triggered and colorized. The scaling of those frames is set by the scene it uses and not by a global setting. The global setting is only applied when upscaling 128x32 projects.

lucky01 commented:
In my dmddevice.ini I have the following settings
style.default.dotsize = 0.85
style.default.dotrounding = 0.85
style.default.brightness = 0.95
style.default.dotglow = 0.3
style.default.backglow = 0.4
style.default.dotsharpness = 0.8

When I compile you latest code my virtal dmd looks like this 
Any idea ?
![virtDMD](https://github.com/freezy/dmd-extensions/assets/14914346/4b426071-407a-41db-acbb-266fc1b0d526)

lucky01 commented:
If I comment the settings out , it looks o.k.

freezy commented:
Hmm that's weird. Which game/colorization is that?

lucky01 commented:
Scared Stiff

lucky01 commented:
... but it happens on all tables no matter colorized or not, I just recognized that it only happens on the first table I start with PinMame Setup64.exe. On the second start all looks fine.

jsm174 commented:
@lucky01 - after this PR gets merged, will the changes to [AlphaNumeric.cs](https://github.com/freezy/dmd-extensions/pull/374/commits/5cd3ec3190112a7529a703da800995de1f3753ad) get retrofitted back into PinMAME's https://github.com/vpinball/pinmame/blob/master/ext/dmddevice/usbalphanumeric.h?

I was looking into Rock with the 2x20 display, and saw it getting truncated when converted to 128x32.

Three other observations:
- what changes would be needed in [dmddevice.cpp](https://github.com/vpinball/pinmame/blob/master/src/wpc/dmddevice.cpp#L346) to handle the two new layouts: `__1x7Num_1x16Alpha_1x16Num`, `__1x16Alpha_1x16Num_1x7Num_1x4Num`?
- I noticed all the `SmoothDigitCorners` calls were moved out of the inner loop, except for `Render2x16Alpha`, guessing this was missed?
- Double `clear()` calls here: https://github.com/freezy/dmd-extensions/blob/5cd3ec3190112a7529a703da800995de1f3753ad/LibDmd/DmdDevice/AlphaNumeric.cs#L476-L478

lucky01 commented:
> @lucky01 - after this PR gets merged, will the changes to [AlphaNumeric.cs](https://github.com/freezy/dmd-extensions/pull/374/commits/5cd3ec3190112a7529a703da800995de1f3753ad) get retrofitted back into PinMAME's https://github.com/vpinball/pinmame/blob/master/ext/dmddevice/usbalphanumeric.h?
> 

I attached the file I used for my dmddevice.dll

> I was looking into Rock with the 2x20 display, and saw it getting truncated when converted to 128x32.
> 
> Three other observations:
> 
> * what changes would be needed in [dmddevice.cpp](https://github.com/vpinball/pinmame/blob/master/src/wpc/dmddevice.cpp#L346) to handle the two new layouts: `__1x7Num_1x16Alpha_1x16Num`, `__1x16Alpha_1x16Num_1x7Num_1x4Num`?

I added those two lines to my dmddevice.dll 

	if ((strncasecmp(gameName, "rvrbt_", 6) == 0)) layout = __1x16Alpha_1x16Num_1x7Num_1x4Num;
	if ((strncasecmp(gameName, "polic_", 6) == 0)) layout = __1x7Num_1x16Alpha_1x16Num;

layout_t looks like this

typedef enum { 
	None,
	__2x16Alpha, 
	__2x20Alpha, 
	__2x7Alpha_2x7Num, 
	__2x7Alpha_2x7Num_4x1Num, 
	__2x7Num_2x7Num_4x1Num, 
	__2x7Num_2x7Num_10x1Num, 
	__2x7Num_2x7Num_4x1Num_gen7, 
	__2x7Num10_2x7Num10_4x1Num,
	__2x6Num_2x6Num_4x1Num,
	__2x6Num10_2x6Num10_4x1Num,
	__4x7Num10,
	__6x4Num_4x1Num,
	__2x7Num_4x1Num_1x16Alpha,
	__1x16Alpha_1x16Num_1x7Num,
	__1x7Num_1x16Alpha_1x16Num,
	__1x16Alpha_1x16Num_1x7Num_1x4Num
} layout_t;


> * I noticed all the `SmoothDigitCorners` calls were moved out of the inner loop, except for `Render2x16Alpha`, guessing this was missed?

Seems I missed that.


[usbalphanumeric.zip](https://github.com/freezy/dmd-extensions/files/11890591/usbalphanumeric.zip)

jsm174 commented:
Great. Thanks. Since the zip is attached, you want me to get that in pinmame or do you want to do it?

lucky01 commented:
You can do it

Am Mi., 28. Juni 2023 um 13:02 Uhr schrieb jsm174 ***@***.***
>:

> Great. Thanks. Since the zip is attached, you want me to get that in
> pinmame or do you want to do it?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/freezy/dmd-extensions/pull/374#issuecomment-1611199617>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADRZGKSWQ7I5PSKXPC4WJTLXNQFLRANCNFSM6AAAAAAYRDBRXM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

lucky01 commented:
I don´t understand the change in DMDFrame.cs Is there a case where BitLength is 3 ?
Even if the number of rawframes is 3 for WPC the Bitlength of the frame should still be 2.
The same for AlvinG, Capcom etc where the number of Rawframes if 4 or 5 (I don´t rember exactly) but the BitDepth is also 2.
The BitDepth should only be determined by where it is coming in (Render2 or Render4)  and is independend from the number of rawframes.

freezy commented:
Yes, that's true, and I'll clean that up.

lucky01 commented:
The latest changes seem to fix the blurry output. 
I started to do some testing and would prefer to remove

if (!File.Exists(dllPath)) {
	Logger.Error("[plugin] Ignoring plugin defined at " + dllPath + ", file does not exist.");
	return false;
}

from ColorizationPlugin.cs

It works the same way but it also searches the plugin dll in different DLL locations.

freezy commented:
Hang on, you're saying that [8bb965f](https://github.com/freezy/dmd-extensions/pull/374/commits/8bb965fee30e5bbecce36478a47c25853664ce1a) fixed it?

> It works the same way but it also searches the plugin dll in different DLL locations.

You mean, if only the filename and not the entire path is provided, keeping this code would throw the error even though `LoadLibrary()` later would work?

lucky01 commented:
> Hang on, you're saying that [8bb965f](https://github.com/freezy/dmd-extensions/pull/374/commits/8bb965fee30e5bbecce36478a47c25853664ce1a) fixed it?
> 
Seems so !
> > It works the same way but it also searches the plugin dll in different DLL locations.
> 
> You mean, if only the filename and not the entire path is provided, keeping this code would throw the error even though `LoadLibrary()` later would work?

Yes !

lucky01 commented:
I recognized that your dedupe check in 
public void NextFrame(ColoredFrame frame)
does not work properly.
e.g. if you use the vni colorization for cc_13 the rom check bar which is normally running from left to right is not updating

freezy commented:
Did you test this? `File.Exists()` should use the same current working dir as `NativeDllLoad.LoadLibrary()`, so both should work.

> e.g. if you use the vni colorization for cc_13 the rom check bar which is normally running from left to right is not updating

Okay, I'll try to test this. I don't have `cc_13` as VNI, though.

lucky01 commented:
https://vpuniverse.com/files/file/10073-cactus-canyon-colorization-v-10-64-colors/

lucky01 commented:
> Did you test this? `File.Exists()` should use the same current working dir as `NativeDllLoad.LoadLibrary()`, so both should work.

Yes, but NativeDllLoad also check additional directories like exe path etc.

freezy commented:
I see. Don't you think it's better for the user when they see a concise message? The problem I see is that if they put a path that doesn't exist that `NativeDllLoad.LoadLibrary()` cannot find either, the message is "Error loading plugin at <path>", whereas if they put the wrong DLL (like, your current one) at the *correct* path, it's also that same message.

So it's a lot harder to tell them where to look when it doesn't work and they provide a log.

lucky01 commented:
The wrong DLL (my current one) reports
[plugin] Failed to open colorizer plugin. 

whereas if it is not found it reports 
[plugin] Error loading plugin at pin2color64.dll. 

Maybe change the first message to Logger.Info($"[plugin] Failed to open colorizer plugin {GetName()} v{GetVersion()}"); ?

freezy commented:
Fair enough. Pushed the changes and the de-dupe fix.

lucky01 commented:
What about adding the version and name to the failed to open message ?

lucky01 commented:
Could you add something like if we get a
[plugin] Error loading plugin at pin2color64.dll.
that is looks for the DLL next to vpm ?
Maybe with somehing like PathUtil.GetVpmPath ?

freezy commented:
Hmm, yeah... Although if people put the absolute path, there shouldn't be any issues, right?

freezy commented:
Try this..

lucky01 commented:
> Hmm, yeah... Although if people put the absolute path, there shouldn't be any issues, right?

Yes, that is how I currently tested and looks good so far. The latest changes seem to work as expected but unfortunately not much time to continue testing today. Maybe tomorrow.

freezy commented:
Okay. Public testing this weekend would be nice.

lucky01 commented:
Did you make a new build with the latest changes (VPM folder) ?

lucky01 commented:
I have some reliable testers which I prefer instead of public testing. Too many testers without basic knowledge about the system.

freezy commented:
> Did you make a new build with the latest changes (VPM folder) ?

Yes, new builds are automatically available. Click on the green check mark beside the commit SHA, choose "Details", which brings you to AppVeyor. There, choose the platform and select the "Artifacts" tab on the top right, where you can download them.

> I have some reliable testers which I prefer instead of public testing.

Did they already start testing? Any ETA for a public release?

lucky01 commented:
> Did they already start testing? Any ETA for a public release?

Yes ! There is already a issue with the simple palette and pin2dmd which I need to debug.

lucky01 commented:
> Did they already start testing? Any ETA for a public release?

When all issues they find are fixed.

freezy commented:
I suppose you're referring to [this line](https://github.com/freezy/dmd-extensions/blob/plugins/LibDmd/Output/Pin2Dmd/Pin2Dmd.cs#L146). Here, we assume that `ColoredFrame frame` is 2-bit (if that's not the case, then that might be the problem).

The actual planes are only stored in `RawFrame`, so `DmdFrame.BitPlanes` always computes the planes from `Data`, based on `BitLength`. If `ColoredFrame frame` is indeed 4-bit, then the question is why `RenderColoredGray2()` was executed instead of `RenderColoredGray4()`.

Does that make sense?

lucky01 commented:
That has nothing to do with the plugin or RGB24. It is happening even with internal handling of a simple 4 color palette.

freezy commented:
Okay.. Any way I can help debugging?

lucky01 commented:
If I change in coloredFrame.cs

public DmdFrame ConvertToGray(params byte[] mapping) 

to

public DmdFrame ConvertToGray4(params byte[] mapping)
{
	return new DmdFrame(Dimensions, FrameUtil.ConvertGrayToGray(Data, mapping), 4);
}

and change also in pin2dmd.cs

public void RenderColoredGray2(ColoredFrame frame)
{
	SetPalette(frame.Palette);

	// send frame buffer to device
	RenderGray4(frame.ConvertToGray4(0x0, 0x1, 0x4, 0xf));
}

It seems to work from what I can see.

lucky01 commented:
The number items in a mapping does not relate mathematically to the bitlength. You could only derive it from the biggest value in the mapping but that is not necessary.

freezy commented:
Pushed.

lucky01 commented:
I found another issue. The data is not forwarded to the plugin when colorization is disabled in pinmame.
For dumping it should forward always forward the data but only receive and process RGB24 data when colorization is active.

freezy commented:
Okay, I'll have a look later tonight.

freezy commented:
Try this. I've only tested it briefly though.

lucky01 commented:
Thanks ! I will test this soon but this weekend is very occupied with private stuff.
From my testers I have a issue reported that puptriggers are not working with the 32bit version but are with the 64bit version (latest dmddevicePUP.dll from https://www.nailbuster.com/wikipinup/doku.php?id=update_baller_may_2023). If you set showdmd in the dmddevicepup.ini you can see that there is no output if a colorization is active (e.g. cc_13) with the 32bit version, but there is with the 64bit version.

freezy commented:
Okay, thanks, I'll try debugging this then.

freezy commented:
Okay, I've actually talked to @Nailbuster about this a few weeks ago, and the behavior should be (and is) the following:

- If `dmddevicePUP.dll` supports consuming triggers, AND the colorizer supports emitting triggers, disable frame passthrough to pinUP, and just send triggers.
- Otherwise, enable frame passthrough (without triggers).

I think the reason why your testers had an image at 64-bit and not in 32-bit is because they only have (like me) the 32-bit version of `pin2color.dll`. So for 32-bit, it would load the coloring plugin which supports pup triggers, thus no DMD output, while on 64-bit, they would fall back to native VNI colorization, which doesn't support pup triggers, and thus enables DMD passthrough. 

I fixed another, unrelated bug, in PinUp while testing though.

lucky01 commented:
He has both 32bit and 64Bit dmddevicepup.dll and dmddevicepup64.dllAm 08.07.2023 um 12:56 schrieb freezy ***@***.***>:﻿
Okay, I've actually talked to @nailbuster about this a few weeks ago, and the behavior should be (and is) the following:

If dmddevicePUP.dll supports consuming triggers, AND the colorizer supports emitting triggers, disable frame passthrough to pinUP, and just send triggers.
Otherwise, enable frame passthrough (without triggers).

I think the reason why your testers had an image at 64-bit and not in 32-bit is because they only have (like me) the 32-bit version of pin2color.dll. So for 32-bit, it would load the coloring plugin which supports pup triggers, thus no DMD output, while on 64-bit, they would fall back to native VNI colorization, which doesn't support pup triggers, and thus enables DMD passthrough.
I fixed another, unrelated bug, in PinUp while testing though.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

lucky01 commented:
He uses the 32bit version which supports Pup triggers Am 08.07.2023 um 12:56 schrieb freezy ***@***.***>:﻿
Okay, I've actually talked to @nailbuster about this a few weeks ago, and the behavior should be (and is) the following:

If dmddevicePUP.dll supports consuming triggers, AND the colorizer supports emitting triggers, disable frame passthrough to pinUP, and just send triggers.
Otherwise, enable frame passthrough (without triggers).

I think the reason why your testers had an image at 64-bit and not in 32-bit is because they only have (like me) the 32-bit version of pin2color.dll. So for 32-bit, it would load the coloring plugin which supports pup triggers, thus no DMD output, while on 64-bit, they would fall back to native VNI colorization, which doesn't support pup triggers, and thus enables DMD passthrough.
I fixed another, unrelated bug, in PinUp while testing though.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

freezy commented:
> He has both 32bit and 64Bit dmddevicepup.dll and dmddevicepup64.dll

It's not about dmddevicePUP but pin2color

> He uses the 32bit version which supports Pup triggers 

Yes, see explanations above.

lucky01 commented:
I had a look and the problem is that the c++ bool gets misinterpreted by c#
To fix this you need to change
private delegate bool HasEventsPtr(); 
to
private delegate byte HasEventsPtr();
and then assign
_hasEvents = System.Convert.ToBoolean(_hasEventsPtr());

The same for 
private delegate byte OpenPtr();
and
// open plugin
if (!(System.Convert.ToBoolean(_open()))) {

freezy commented:
Pushed.

lucky01 commented:
In general it is working now but still the 32bit version has problems.
It seems to miss some triggers e.g.
Baywatch 1.0.5 (192 x 64)
.pal/.vni color files, 32bit some pups trigger, many do not (shark, characters). With 64 bit all triggers work
He reverted back to my version which works without problems.
Is there a dupe check for pinup active or is it getting frames at 60Hz ?

freezy commented:
Which triggers are you referring to? The ones triggered by the plugin, or the ones triggered in PinUP through frames?

There are so many variables that it's hard to tell anything without a log. If you could provide that, or more detailed instruction how to reproduce, I might be able to help. Usually, when there's a difference between 32 and 64 bit, it's because of a third party dependency that is either available or not on one platform. But again, hard to tell without a log.

Frames are de-duped when forwarded directly to PUP, in both 32- and 64-bit. Frames to the plugin that send the triggers are not de-duped.

lucky01 commented:
So far there is no colorization project I know of which uses the new mode and triggers pup directly, so I´m talking about the frames forwarded to the dmddevicePUP.dll which then trigger the scenes. To reproduce just install the table with the puppack and start it with a 32bit and 64bit vpx/pup installation. I received the same issue report when I deduped the frames going to dmddevicepup.dll and only got it fixed by removing the dedupe code again. It seems that to make the 32bit version work it needs to receive the triggering frame multiple times (60hz) to trigger the pup scenes reliably.

freezy commented:
@nailbuster is this something that could be fixed on pinup side?

lucky01 commented:
Here I tried the deduplication for pup 
https://github.com/lucky01/dmd-extensions/commit/c15f3dc2d9b3e43266dfd5a984f5ae68c72543de
I had to revert it to make the triggers work.

freezy commented:
Okay, but we can agree that there is not really a need to duplicate frames for event matching, right? For the colorizer that's different because it might generate additional frames. @nailbuster could this be a fix in pup?

nailbuster commented:
pupdll frame matching just works on each individual frame sent,  you could send 1 per second or 120 per second.  

to debug,  if you want to see what pupdll is getting from dmdext.  In vpinmame folder set showdmd=1 in the dmddevicepup.ini

from memory i remember only getting 'changed' frames which i thought was great as it meant i didn't need to worry checking for dupes on the pup side.

lucky01 commented:
> to debug, if you want to see what pupdll is getting from dmdext. In vpinmame folder set showdmd=1 in the dmddevicepup.ini you should see a small 128x32 window at position 0,0

I already checked that and the frames are forwarded correctly, however it seems that the matching does not work reliably with the 32bit version.

> from memory i remember only getting 'changed' frames which i thought was great as it meant i didn't need to worry checking for dupes on the pup side.

No in previous versions you received the frames at 60Hz without dupe check.

freezy commented:
I can't reproduce the FlexDMD issue you have on VPU (it's down, so posting here). If you pull latest and compile it with the `TEST_WITHOUT_PIN2DMD` compiler flag, it'll just write the frame size to a `pin2dmd.txt` file for PIN2DMDs. That's how I'm testing without the device.

I'm getting frame data for both PIN2DMD and virtual DMD. Can you reproduce with that? If so, can you share your `DmdDevice.ini`?

TBH, I don't think it's fair to stall because of this. It's literally and edge case of an edge case that has nothing to do with coloring.

lucky01 commented:
Sure ! Can do next week ! The more critical issue is PUP triggers not working correctly on 32bit compile. Any updates ?

lucky01 commented:
I´m not stalling anything here I just want to make sure it is working at least as good as my current version and does not contain issues that were already solved there.

lucky01 commented:
> I'm getting frame data for both PIN2DMD and virtual DMD. Can you reproduce with that?

I can reproduce that however when I start FlexDMDUI, go to the designer tab and click on run you can see that with virtual set to false it only sends 2 packets while with set to true it constantly sends packets when the display changes. I assume without  virtual it only sends the init frames but not the display content.

lucky01 commented:
When you exit flexdmdUI it sends another 3 frames for deinit.

freezy commented:
Thanks, now I see it. Looking into it!

freezy commented:
Try this. I haven't tested everything because I'm on vacation and it's difficult to launch actual games through remote desktop, but the missing output should be fixed.

I'll look into pinup when I'm back home, but this might got fixed as a side effect as well. If you could verify that'd be awesome.

lucky01 commented:
> I'll look into pinup when I'm back home, but this might got fixed as a side effect as well. If you could verify that'd be awesome.

I don´t have pinup installed but will ask someone to test. Enjoy your vacation

lucky01 commented:
FlexdmdUI works fine now but

> Baywatch 1.0.5 (192 x 64)
> .pal/.vni color files, 32bit some pups trigger, many do not

still the same problem !

freezy commented:
@lucky01 any hint of a trigger in Baywatch that I could test? I have Pup set up now, but I'm kinda clueless what to look for.

lucky01 commented:
I can only forward the information I receive, because I don´t have pup installed. 
It seems that the 32bit does not trigger at all or rarely while the 64bit version does work quite well.
My version is stil working best in both x86 and x64. Maybe just play the table and see the difference.
